ocaml/testsuite/tests/typing-modes/currying.ml

```diff
@@ -458,6 +458,8 @@ val f : ('a -> 'b -> 'c) -> unique_ 'a -> 'b -> 'c = <fun>
 let f (g @ unique) x =
   g x x [@nontail]
 [%%expect{|
+val f : unique_ ('a -> 'a -> 'b) -> ('a -> 'b) = <fun>
+|}, Principal{|
 val f : unique_ ('a -> 'a -> 'b) -> 'a -> 'b = <fun>
 |}]
```

Comment:
I have no idea what's going on here

Reply:
I think the observed behavior is correct.

The second type really means `unique_ ('a -> 'a -> 'b) -> once_ ('a -> 'b)`. The `once_` is necessary to deal with the possibility of partial application: `f x` is actually a closure with a pointer to the unique `x`, and such things must be `once`. The first type, with its seemingly-redundant parens, means just what it says. That makes sense, because the function can mode-cross uniqueness, and so we don't actually need the `once`. Yet the reason we want the `once` in the second type is because of principality: we only infer that `g` is a function by its usage, which is not principal. If you put a type signature on `g`, then I think you'd only get the second type.

Bottom line: all is well here.